src/app/product/[productId]/page.tsx

```diff
@@ -0,0 +1,147 @@
+"use client";
+import { Formik, Form, Field, ErrorMessage } from "formik";
+import axios from "axios";
+import * as Yup from "yup";
+import { Card, CardContent, CardFooter, CardHeader, CardTitle} from "@/components/ui/card";
+import { Button } from "@/components/ui/button";
+import { Input } from "@/components/ui/input";
+import { Label } from "@/components/ui/label";
+import useRequests from "../../service/request";
+import { Separator } from "@/components/ui/separator";
+import { useToast } from "@/components/ui/use-toast";
+import { useRouter } from "next/navigation";
+import { useEffect, useState } from "react";
+import { useParams } from "next/navigation";
+
+interface Product {
+  name: string;
+  description: string;
+  price: number | null;
+}
+
+const validationSchema = Yup.object().shape({
+  name: Yup.string().required("Name is required"),
+  description: Yup.string().required("Description is required"),
+  price: Yup.number()
+    .typeError("Price must be a number")
+    .required("Price is required")
+    .positive("Price must be positive"),
+});
+
+const EditProduct = () => {
+  const { productId } = useParams();
+  const { toast } = useToast();
+  const router = useRouter();
+  const endpoints = useRequests();
+  const [initialValues, setInitialValues] = useState({
+    name: "",
+    description: "",
+    price: null,
+  });
+
+  useEffect(() => {
+    fetchProduct();
+  }, [productId]);
+
+  const fetchProduct = async () => {
+    try {
+      const response = await axios.get(`${endpoints.getProductById}/${productId}`);
+      setInitialValues(response.data);
+    } catch (error) {
+      console.error(error);
+    }
+  };
+
+  const onSubmit = async (values:Product, { resetForm }:{ resetForm: () => void }) => {
+    try {
+      const response = await axios.put(`${endpoints.updateProductById}/${productId}`, values);
```

Comment:
there are lot of caching options in next js with fetch, you can look through them. Also look at how server actions/function work